### PR TITLE
METRON-569: Enrichment topology duplicates messages

### PR DESCRIPTION
**Pull request by merrimanr**
## Contributor Comments
This issue was discovered a while ago but a PR with a fix was never accepted.  In summary, the JoinBolt is acking the last tuple of a split message instead of the tuple containing the original message sent directly from the SplitBolt.  This is a problem because the tuple with the original message is anchored in the SplitBolt and the other tuples that go to enrichment bolts are not.

The major change in this PR involves storing the tuples in the cache instead of just the message.  This way the tuple from the SplitBolt can be retrieved and acked once all the parts have been received in the JoinBolt.  

A couple other minor changes are also included.  I refactored the JoinBoltTest to make it easier to follow and maintain.  If anyone thinks this is worse than what we currently have I can revert that.  I also bumped up the default cache sizes for the JoinBolts from 10,000 to 100,000.  The low initial cache size was causing failures in full dev if I let it run for a while.

## Pull Request Checklist

Thank you for submitting a contribution to Apache Metron.  
Please refer to our [Development Guidelines](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=61332235) for the complete guide to follow for contributions.  
Please refer also to our [Build Verification Guidelines](https://cwiki.apache.org/confluence/display/METRON/Verifying+Builds?show-miniview) for complete smoke testing guides.  

In order to streamline the review of the contribution we ask you follow these guidelines and ask you to double check the following:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? If not one needs to be created at [Metron Jira](https://issues.apache.org/jira/browse/METRON/?selectedTab=com.atlassian.jira.jira-projects-plugin:summary-panel). 
- [ ] Does your PR title start with METRON-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.
- [ ] Has your PR been rebased against the latest commit within the target branch (typically master)?

### For code changes:
- [x] Have you included steps to reproduce the behavior or problem that is being changed or addressed?
- [x] Have you included steps or a guide to how the change may be verified and tested manually?
- [x] Have you ensured that the full suite of tests and checks have been executed in the root incubating-metron folder via:
  ```
  mvn -q clean integration-test install && build_utils/verify_licenses.sh 
  ```

- [x] Have you written or updated unit tests and or integration tests to verify your changes?
- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [x] Have you verified the basic functionality of the build by building and running locally with Vagrant full-dev environment or the equivalent?

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered by building and verifying the site-book? If not then run the following commands and the verify changes via `site-book/target/site/index.html`:

  ```
  cd site-book
  mvn site
  ```

#### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.
It is also recommended that [travis-ci](https://travis-ci.org) is set up for your personal repository such that your branches are built there before submitting a pull request.

### CODE REVIEW

File: metron-platform/metron-enrichment/src/main/java/org/apache/metron/enrichment/bolt/JoinBolt.java [hunk 1]
```diff
@@ -141,11 +141,12 @@ public void execute(Tuple tuple) {
...
                       , tuple
                       , new Values( key
-                                  , joinMessages(streamMessageMap)
+                                  , joinMessages(streamMessageMap, this.messageGetStrategy)
                                   )
                       );
         cache.invalidate(key);
-        collector.ack(tuple);
+        Tuple messageTuple = streamMessageMap.get("message:");
+        collector.ack(messageTuple);
```
Comment (cestella) [inline metron-platform/metron-enrichment/src/main/java/org/apache/metron/enrichment/bolt/JoinBolt.java]: Do we need to ack every tuple in `streamMessageMap`?

Comment (merrimanr) [inline metron-platform/metron-enrichment/src/main/java/org/apache/metron/enrichment/bolt/JoinBolt.java]: I don't believe so.  The tuple that comes straight from the SplitBolt (the "message" stream) is the only tuple that is anchored.  Tuples that go to EnrichmentBolts are not anchored.

Comment (cestella) [inline metron-platform/metron-enrichment/src/main/java/org/apache/metron/enrichment/bolt/JoinBolt.java]: Sounds good!

**Comment by cestella** [comment]
+1 by inspection, great catch!

**Pull request closed without merging**